File: cpp/robot-simulator/robot_simulator.cpp
#include <iostream>
#include "robot_simulator.h"

namespace robot_simulator {

    Robot::Robot()
    {
        this->position = {0, 0};
        this->bearing = Bearing::NORTH;
    }

    Robot::Robot(std::pair<int, int> position, Bearing bearing)
    {
        this->position = position;
        this->bearing = bearing;
    }

    void Robot::turn_left()
    {
        switch (this->bearing)
        {
            case Bearing::NORTH:
                this->bearing = WEST;
                break;
            case Bearing::EAST:
                this->bearing = NORTH;
                break;
            case Bearing::SOUTH:
                this->bearing = EAST;
                break;
            case Bearing::WEST:
                this->bearing = SOUTH;
        }
    }

    void Robot::turn_right()
    {
        switch (this->bearing)
        {
            case Bearing::NORTH:
                this->bearing = EAST;
                break;
            case Bearing::EAST:
                this->bearing = SOUTH;
                break;
            case Bearing::SOUTH:
                this->bearing = WEST;
                break;
            case Bearing::WEST:
                this->bearing = NORTH;
        }
    }

    void Robot::advance()
    {
        switch (this->bearing)
        {
            case Bearing::NORTH:
                this->position.second++;
                break;
            case Bearing::EAST:
                this->position.first++;
                break;
            case Bearing::SOUTH:
                this->position.second--;
                break;
            case Bearing::WEST:
                this->position.first--;
        }
    }

    void Robot::execute_sequence(std::string commands)
    {
        for (auto command : commands)
        {
            switch (command)
            {
                case 'A':
                    this->advance();
                    break;
                case 'R':
                    this->turn_right();
                    break;
                case 'L':
                    this->turn_left();
                    break;
                default:
                    break;
            }
        }
    }

    std::pair<int, int> Robot::get_position()
    {
        return this->position;
    }

    Bearing Robot::get_bearing()
    {
        return this->bearing;
    }

}  // namespace robot_simulator

File: python/robot-simulator/robot_simulator.py
class Direction:
	def __init__(self, xOffset, yOffset):
		self.xOffset = xOffset
		self.yOffset = yOffset

	def __eq__(self, other) :
		return type(self) == type(other) and self.xOffset == other.xOffset and self.yOffset == other.yOffset

	def __repr__(self):
		return "({}, {})".format(self.xOffset, self.yOffset)

	def __hash__(self):
		return hash((self.xOffset, self.yOffset))


EAST = Direction(1, 0)
NORTH = Direction(0, 1)
WEST = Direction(-1, 0)
SOUTH = Direction(0, -1)


class Robot:

	right_conversions = {
		NORTH: EAST,
		EAST: SOUTH,
		SOUTH: WEST,
		WEST: NORTH
	}

	left_conversions = {
		NORTH: WEST, 
		EAST: NORTH,
		SOUTH: EAST,
		WEST: SOUTH
	}

	def __init__(self, direction=NORTH, x=0, y=0):
		self.coordinates = (x, y)
		self.direction = direction
		self.functions = {
			"A": self.advance,
			"R": self.turn_right,
			"L": self.turn_left
		}

	def advance(self):
		self.coordinates = (self.coordinates[0] + self.direction.xOffset, self.coordinates[1] + self.direction.yOffset)

	def turn_right(self):
		self.direction = self.right_conversions[self.direction]

	def turn_left(self):
		self.direction = self.left_conversions[self.direction]

	def move(self, commands: str):
		for command in commands:
			self.functions[command]()
